Index: linux-4.4.198/arch/mips/ralink/mt7621.c
===================================================================
--- linux-4.4.198.orig/arch/mips/ralink/mt7621.c
+++ linux-4.4.198/arch/mips/ralink/mt7621.c
@@ -149,6 +149,7 @@ void __init ralink_clk_init(void)
 	u8 clk_sel, xtal_mode, mc_prediv_sel;
 	u64 cpu_clk;
 	u32 xtal_clk;
+	int i;
 
 	xtal_mode = (rt_sysc_r32(SYSC_REG_SYSTEM_CONFIG0) >> XTAL_MODE_SEL_SHIFT)
 		    & XTAL_MODE_SEL_MASK;
@@ -169,6 +170,10 @@ void __init ralink_clk_init(void)
 		break;
 	case 1:
 		mc_cpll = rt_memc_r32(DRAMC_REG_MPLL18);
+		mc_cpll &= ~(0x7ff);
+		mc_cpll |= (0x312);
+		rt_memc_w32(mc_cpll, DRAMC_REG_MPLL18);
+		for (i=0; i<1024;i++);
 		mc_fb = (mc_cpll >> CPLL_FBDIV_SHIFT) & CPLL_FBDIV_MASK;
 		mc_prediv_sel = (mc_cpll >> CPLL_PREDIV_SHIFT)
 				& CPLL_PREDIV_MASK;
@@ -204,6 +209,7 @@ void __init ralink_clk_init(void)
 	cpu_rate = cpu_clk & 0xffffffff;
 
 	mips_hpt_frequency = cpu_rate / 2;
+	pr_emerg("CPU Clock: %dMHz\n", cpu_rate/1000000);
 }
 
 void __init ralink_of_remap(void)
